File: best-time-to-buy-and-sell-stocks.ts
function maxProfit(prices: number[], bought: number | null = null, sold: number | null = null): number {
  //     let minPrice = prices[0];
  //   let profit = 0;
  
  //   for (let currentPrice of prices) {
  //     if (currentPrice > minPrice) {
  //       profit = Math.max(profit, currentPrice - minPrice);
  //     } else {
  //       minPrice = currentPrice;
  //     }
  //   }
  //   return profit;
      
      // at any given point you can either buy or sell
  
      // base cases
      if((bought!==null) && (sold!==null)) {
          return sold - bought;
      }
      
      if (!prices || prices.length === 0){
          return 0
      }
  
      const sells = [];
      if (bought === null) {
          // must buy or hold but cannot sell
          sells.push(
              maxProfit(prices.slice(1), bought, null) // hold scenario
          )
          sells.push(
              maxProfit(prices.slice(1), prices[0], null) // buy scenario
          )
      } else {
          // can sell or hold
          sells.push(
              maxProfit(prices.slice(1), bought, null) // hold scenario
          )
          sells.push(
              maxProfit(prices.slice(1), bought, prices[0]) // sell scenario
          )
      }
      //console.log(prices, sells)
      return Math.max(...sells);
  };

File: restart_prep/best_time_to_buy_and_sell_stocks.cpp
#include <algorithm>
#include <utility>
#include <vector>

/*
 *You are given an array prices where prices[i] is the price of a given stock on the ith day.

You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.
 *
 * */

/*
 * Pretend to make a sale everyday, tracking the lowest buying price so far.
 **/
class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
		int maxProfit = 0;
		if (prices.size() == 0)
			return maxProfit;

		int lowestBuySoFar = prices[0];
		for (int i = 0; i < prices.size(); i++) {
			int currPrice = prices[i];
			lowestBuySoFar = std::min(currPrice, lowestBuySoFar);
			int ifSoldToday = currPrice - lowestBuySoFar;
			maxProfit = std::max(maxProfit, ifSoldToday);
		}
		return maxProfit;
    }
};
